src/sections/layer_and_mask_information_section/mod.rs

```diff
@@ -110,67 +111,54 @@ impl LayerAndMaskInformationSection {
         ) = LayerAndMaskInformationSection::read_layer_records(&mut cursor, layer_count)?;
 
         LayerAndMaskInformationSection::decode_layers(
-            &mut cursor,
             layer_records,
-            layer_count as usize,
             group_count,
             (psd_width, psd_height),
         )
     }
 
     fn decode_layers(
-        cursor: &mut PsdCursor,
         layer_records: Vec<(LayerRecord, LayerChannels)>,
-        layer_count: usize,
         group_count: usize,
         psd_size: (u32, u32),
     ) -> Result<LayerAndMaskInformationSection, Error> {
-        let mut layers = NamedItems::with_capacity(layer_count);
+        let mut layers = NamedItems::with_capacity(layer_records.len());
         let mut groups: NamedItems<PsdGroup> = NamedItems::with_capacity(group_count);
 
-        let mut levels: Vec<Vec<Frame>> = vec![];
-        levels.push(vec![Frame {
+        // Create stack with root-level
+        let mut stack: Vec<Frame> = vec![Frame {
             start_idx: 0,
             name: String::from("root"),
             group_id: 0,
             parent_group_id: 0,
-        }]);
+        }];
+
+        // Viewed group counter
         let mut already_viewed = 0;
-        let mut nested_level = 0;
-        // Read each layer's channel image data
-        for (idx, item) in layer_records.into_iter().enumerate() {
-            let layer_record = item.0;
-            let channels = item.1;
 
-            let frame = levels.get_mut(nested_level).unwrap();
-            let parent_group_id = frame.last().unwrap().group_id;
+        // Read each layer's channel image data
+        for (layer_record, channels) in layer_records.into_iter() {
+            // get current group from stack
+            let current_group_id = stack.last().unwrap().group_id;
 
             match layer_record.divider_type {
                 // open the folder
                 Some(GroupDivider::CloseFolder) | Some(GroupDivider::OpenFolder) => {
```

Comment:
Why is this both close and open folder?

Reply:
So, I thought that `OpenFolder` is open tag, `CloseFolder` is close tag. It's wrong.
They both are open tags, the difference is, folder is closed or not in layers list.

`CloseFolder`
![image](https://user-images.githubusercontent.com/15220500/79041982-9cbf9e00-7bfc-11ea-8c8d-783d00b54f2a.png)

`OpenFolder`
![image](https://user-images.githubusercontent.com/15220500/79041993-ae08aa80-7bfc-11ea-9486-9fafc481270a.png)

Reply:
Ohhhhhhh - nice!

Mind just commenting that on their enum definitions?